src/erlcloud_cognito.erl

```diff
@@ -0,0 +1,906 @@
+                          Session, Cfg) ->
+    respond_to_auth_challenge(ClientId, ChallengeName, ChallengeResponses,
+                              Session, #{}, #{}, #{}, Cfg).
+
+-spec respond_to_auth_challenge(binary(), binary(), maps:map(), binary(),
+                                maps:map(), maps:map(), maps:map(),
+                                aws_config()) ->
+    {ok, map()} | {error, any()}.
+respond_to_auth_challenge(ClientId, ChallengeName, ChallengeResponses,
+                          Session, AnalyticsMeta, ClientMeta, ContextData, Cfg) ->
+    Body = #{
+        <<"AnalyticsMetadata">>  => AnalyticsMeta,
+        <<"ChallengeName">>      => ChallengeName,
+        <<"ChallengeResponses">> => ChallengeResponses,
+        <<"ClientId">>           => ClientId,
+        <<"ClientMetadata">>     => ClientMeta,
+        <<"Session">>            => Session,
+        <<"UserContextData">>    => ContextData
+    },
+    request(Cfg, "RespondToAuthChallenge", erlcloud_util:filter_empty_map(Body)).
+
+-spec create_identity_provider(binary(), binary(), binary(), map()) ->
+    {ok, map()} | {error, any()}.
+create_identity_provider(UserPoolId, ProviderName, ProviderType,
+                         ProviderDetails) ->
+  create_identity_provider(UserPoolId, ProviderName, ProviderType,
+                           ProviderDetails, #{}).
+
+-spec create_identity_provider(binary(), binary(), binary(), map(), map()) ->
+    {ok, map()} | {error, any()}.
+create_identity_provider(UserPoolId, ProviderName, ProviderType,
+                         ProviderDetails, AttributeMapping) ->
+  create_identity_provider(UserPoolId, ProviderName, ProviderType,
+                           ProviderDetails, AttributeMapping, []).
+
+-spec create_identity_provider(binary(), binary(), binary(),
+                               map(), map(), list()) ->
+    {ok, map()} | {error, any()}.
+create_identity_provider(UserPoolId, ProviderName, ProviderType,
+                         ProviderDetails, AttributeMapping, IdpIdentifiers) ->
+  Config = erlcloud_aws:default_config(),
+  create_identity_provider(UserPoolId, ProviderName, ProviderType,
+                           ProviderDetails, AttributeMapping, IdpIdentifiers, Config).
+
+-spec create_identity_provider(binary(), binary(), binary(), map(), map(),
+                               list(), aws_config()) ->
+    {ok, map()} | {error, any()}.
+create_identity_provider(UserPoolId, ProviderName, ProviderType, ProviderDetails,
+                         AttributeMapping, IdpIdentifiers, Config) ->
+  Body = #{
+    <<"UserPoolId">>       => UserPoolId,
+    <<"ProviderName">>     => ProviderName,
+    <<"ProviderType">>     => ProviderType,
+    <<"ProviderDetails">>  => ProviderDetails,
+    <<"AttributeMapping">> => AttributeMapping,
+    <<"IdpIdentifiers">>   => IdpIdentifiers
+  },
+  request(Config, "CreateIdentityProvider", erlcloud_util:filter_empty_map(Body)).
```

Comment:
such use of erlcloud_util:filter_empty_map() across the file makes the API ignore mandatory specs for AWS API and effectively renders everything as optional leading to unnecessary dev mistakes for bad requests. 

https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_CreateIdentityProvider.html